File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def fun(self,bloomDay,n,m,k,guess):
        flower=0
        bouquet=0

        for i in range(n):
            if(bloomDay[i]<=guess):
                flower+=1
            else:
                bouquet=bouquet+flower//k
                flower=0

        bouquet=bouquet+flower//k

        if(bouquet>=m):
            return True
        else:
            return False

    def minDays(self, bloomDay, m, k):
        n=len(bloomDay)

        if(m*k>n):
            return -1

        low=float('inf')
        high=float('-inf')

        for i in range(n):
            low=min(low,bloomDay[i])
            high=max(high,bloomDay[i])

        res=-1

        while(low<=high):
            guess=(low+high)//2

            if(self.fun(bloomDay,n,m,k,guess)):
                res=guess
                high=guess-1
            else:
                low=guess+1

        return res

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public boolean fun(int[] bloomDay,int n,int m,int k,int guess){
        int flower=0;
        int bouquet=0;

        for(int i=0;i<n;i++){
            if(bloomDay[i]<=guess){
                flower++;
            }
            else{
                bouquet=bouquet+flower/k;
                flower=0;
            }
        }

        bouquet=bouquet+flower/k;

        if(bouquet>=m){
            return true;
        }
        else{
            return false;
        }
    }

    public int minDays(int[] bloomDay, int m, int k) {
        int n=bloomDay.length;

        if((long)m*k>n){
            return -1;
        }

        int low=Integer.MAX_VALUE;
        int high=Integer.MIN_VALUE;

        for(int i=0;i<n;i++){
            low=Math.min(low,bloomDay[i]);
            high=Math.max(high,bloomDay[i]);
        }

        int res=-1;

        while(low<=high){
            int guess=(low+high)/2;

            if(fun(bloomDay,n,m,k,guess)){
                res=guess;
                high=guess-1;
            }
            else{
                low=guess+1;
            }
        }

        return res;
    }
}